go/pkg/client/cas.go

```diff
@@ -1071,56 +1071,62 @@ func (c *Client) FlattenActionOutputs(ctx context.Context, ar *repb.ActionResult
 }
 
 // DownloadDirectory downloads the entire directory of given digest.
-func (c *Client) DownloadDirectory(ctx context.Context, d digest.Digest, execRoot string, cache filemetadata.Cache) (map[string]*TreeOutput, error) {
+// It returns the number of logical and real bytes downloaded, which may be different from sum
+// of sizes of the files due to dedupping and compression.
+func (c *Client) DownloadDirectory(ctx context.Context, d digest.Digest, execRoot string, cache filemetadata.Cache) (map[string]*TreeOutput, int64, int64, error) {
 	dir := &repb.Directory{}
-	if err := c.ReadProto(ctx, d, dir); err != nil {
-		return nil, fmt.Errorf("digest %v cannot be mapped to a directory proto: %v", d, err)
+	_, err := c.ReadProto(ctx, d, dir)
+	if err != nil {
+		return nil, 0, 0, fmt.Errorf("digest %v cannot be mapped to a directory proto: %v", d, err)
 	}
 
 	dirs, err := c.GetDirectoryTree(ctx, d.ToProto())
 	if err != nil {
-		return nil, err
+		return nil, 0, 0, err
 	}
 
 	outputs, err := c.FlattenTree(&repb.Tree{
 		Root:     dir,
 		Children: dirs,
 	}, "")
 	if err != nil {
-		return nil, err
+		return nil, 0, 0, err
 	}
 
-	return outputs, c.downloadOutputs(ctx, outputs, execRoot, cache)
+	lbMoved, rbMoved, err := c.downloadOutputs(ctx, outputs, execRoot, cache)
+	return outputs, lbMoved, rbMoved, err
 }
 
-// DownloadActionOutputs downloads the output files and directories in the given action result.
-func (c *Client) DownloadActionOutputs(ctx context.Context, resPb *repb.ActionResult, execRoot string, cache filemetadata.Cache) error {
+// DownloadActionOutputs downloads the output files and directories in the given action result. It returns the amount of downloaded bytes.
+// It returns the number of logical and real bytes downloaded, which may be different from sum
+// of sizes of the files due to dedupping and compression.
+func (c *Client) DownloadActionOutputs(ctx context.Context, resPb *repb.ActionResult, execRoot string, cache filemetadata.Cache) (logicalBytesMoved, realBytesMoved int64, err error) {
```

Comment:
Suggestion: introduce a DownloadMetadata (or even better, a generalized MovedBytesMetadata or something) with the fields: BytesRequested, RealBytesMoved, LogicalBytesMoved and CachedBytes. There, BytesRequested=sum of all other fields (something is Cached if we don't need to move any bytes to fulfill the request, while the difference between Logical and Real is compression).
That way, you can refactor to just return this MovedBytesMetadata everywhere, and no further changes will be needed if we add/change fields (the functions API will no longer change).
It will also be more consistent across functions (here, sometimes you return one number, sometimes two, sometimes zero). I'd actually be okay with just returning this struct everywhere.
WDYT?

Reply:
I think "no further changes will be needed" is ambitious, but i'll give it a try.